### ISSUE TITLE
Losing inbound email to RDNS lookup failure in Postfix

### ISSUE THREAD

**Issue by mxmfft** [mailcow/mailcow-dockerized#2006]
Postfix rejecting non-spam incoming email due to RDNS lookup errors. 

Around 5 October I stopped getting any incoming email and reviewing logs I noticed every incoming email was failing the RDNS check and getting NOQUEUE: reject from Postfix. This lost email caused some serious issues but I assumed it was a datacentre issue, did a full update OS, update mailcow, reboot and monitor and all looked good. 

A week ago I noticed an email I was expecting had not arrived. I checked the logs and it became apparent that most email was coming in fine but occasional inbound was still getting rejected in the same way. This caused a significant worry as random loss is almost worse than complete loss.

There were some nasty looking SQL format messages in the log so I spun up a brand new VM, did a clean install and migrated to the new server. I have been monitoring logs since. Today I got firm evidence the same is still happening on the new server.

In all cases nslookup was resolving correctly every example I tried.

Postfix log:
NOQUEUE: reject: RCPT from unknown[209.85.221.53]: 450 4.7.1 Client host rejected: cannot find your reverse hostname, [209.85.221.53]; from=<xxx@gmail.com> to=<xxx@xxx> proto=ESMTP helo=<mail-wr1-f53.google.com>

Command line:
nslookup 209.85.221.53
53.221.85.209.in-addr.arpa      name = mail-wr1-f53.google.com.

Most email is coming through fine, this is intermittent.

The unbound logs look clean (getting 1 entry per day) and I cannot find evidence of any other associated problems in the system.

unbound-mailcow_1_de550cfe56fa | 2018-11-14T06:16:06.394404417Z [1542176166] unbound[1:0] info: generate keytag query _ta-4a5c-4f66. NULL IN

I hesitate to raise this here as I know DNS problems can be a real pain and are almost never the fault of an application, but I don't fully understand the operation of unbound so I can't be sure and I'm at the point of having to do "something". I did try asking about this issue on Freenode without success.

Both images are Ubuntu 18.04. VMs are in the DigitalOcean London datacentre, each 1 vCPU and 2GB RAM. The second install was done 7 days ago. I cannot reproduce any other DNS issues outside mailcow.

Please may I ask:

Can this RDNS verification in Postfix be disabled - if so where?
Is there any way inside mailcow-dockerized to increase the number of RDNS tries or boost the timeout?
This isn't greylisting gone wrong is it?
Any ideas of anything else worth trying?

Many thanks for any assistance you can offer.

**Comment by andryyy** [comment]
Around Oct sounds like the DNSSEC key rollover issue. A restart of unbound fixed it. :-)

Do they block or ratelimit DNS queries at DO?

The resolver is only used in mailcow, it does not affect the host system. 

Most of the times it is a shitty provider to block DNS queries or apply ratelimits. But sometimes it happens due to firewalls installed on the host like firewalld or ufw.

Depending on the load another possibility is a oom situation. Running mailcow with SOGo and ClamAV may cause trouble with 2 G of RAM. You should check syslog for the time DNS stopped working. 

It is not a general mailcow issue I can really help you with. :-( You could use another DNSSEC-enabled, non-public resolver on your host system and remove the “dns: “ lines from docker-compose.yml. But you should really check if a container is killed because the host runs oom. 

Edit: I can have a quick look on your machine if you want to. But that only applies to non-commercially used cows.

**Comment by mxmfft** [comment]
Hi Andryyy, thanks very much for your suggestions. 

DO provides pretty good monitoring tools, so ram looks good (~70% used). No sign of unexpected docker image shutdowns or restarts. 

No firewall on the image plus I would expect firewalls to interfere with nslookup if they were active so I don't think that can be the case for an intermittent fault. Rate limiting is a good thought, I'm just running a family mail server that handles maybe 50 messages a day so I know I'm not generating much load but on a shared infrastructure platform its hard to be certain, I can't find any published limits.

It's very kind of you to offer to look at the machine, leave it with me for the moment, I think I'll try swapping systemd-resolve out for something like dnsmasq and see how that goes.

Much appreciate your help, have a great day.

**Comment by andryyy** [comment]
This will probably not help (switching dnsmasq), unbound-mailcow runs totally isolated.

You could try to setup a resolver like unbound/pdns-recursor/whatever on your local node (not anywhere in mailcow) and remove all occurrences of...
```
      dns:
        - ${IPV4_NETWORK:-172.22.1}.254
```
...from docker-compose.yml. Run `docker-compose up -d` afterwards. This would use the systems default resolver. If you don't setup your own resolver on your local node and you use 1.1.1.1/8.8.8.8/etc. instead, a lot of spam filter checks will not work, that's why I asked you to install a local DNS resolver. :-) mailcow would still start and resolve names, but those public nameservers are heavily ratelimited by blacklist providers.

**Comment by mxmfft** [comment]
Thanks. Jumped into this a little more. Turned logging up to 3 in the unbound container and saw a high level of requests, an almost constant stream to xx.highsecure.ru which I believe is rspamd traffic. Absolutely agree this must be cached. 

I noticed from /etc/resolv.conf that systemd is running a local dns service on 127.0.0.53#53.  I tried a naive forward-zone configuration for unbound but that did not work (many SERVEFAIL) so I backed out of that immediately. 

Decided to try removing just postfix dns in docker-compose.yml and let unbound continue handling rspamd and the rest. This seems to be up at the moment. Enabled debug mode on systemd-resolve logging and I can see the postfix lookups.

I'm building off a DigitalOcean pre-built Ubuntu image which comes pre-set with 2 DO dns resolver IPs so fortunately I don't need to chase Google/Cloudflare dns at the moment. The postfix lookup rate is v.low so I will monitor what happens.

Whatever this is it clearly is not a mailcow issue so thank you again for all your help and advice,
